File: 03_liskov_substitution/typescript/bad/index02-bad.ts
class Bird {
  fly (): void {
    console.log('I can fly')
  }
}
class Duck extends Bird {
  quack (): void {
    console.log('I can quack')
  }
}
class Penguin extends Bird {
  fly (): void {
    throw new Error("Cannot fly")
  }
  swim (): void {
    console.log('I can swim')
  }
}

function makeBird (bird: Bird): void {
  bird.fly()
}

const duck = new Duck()
const penguin = new Penguin()

makeBird(duck)
makeBird(penguin)

File: 03_liskov_substitution/dart/bad/index02-bad.dart
class Bird {
  void fly() {
    print('I can fly');
  }
}

class Duck extends Bird {
  void quack() {
    print('I can quack');
  }
}

class Penguin extends Bird {
  void fly() {
    throw Exception('Cannot fly');
  }

  void swim() {
    print('I can swim');
  }
}

makeBird(Bird bird) {
  bird.fly();
}

void main(List<String> args) {
  var duck = Duck();
  var penguin = Penguin();

  makeBird(duck);
  makeBird(penguin);
}
